src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -1962,6 +1962,114 @@ void C2_MacroAssembler::expand_bits_l_v(Register dst, Register src, Register mas
   expand_bits_v(dst, src, mask, /* is_long */ true);
 }
 
+static void float16_to_float_v_slow_path(C2_MacroAssembler& masm, C2GeneralStub<VectorRegister, VectorRegister, uint>& stub) {
+#define __ masm.
+  VectorRegister dst = stub.data<0>();
+  VectorRegister src = stub.data<1>();
+  uint length = stub.data<2>();
+  __ bind(stub.entry());
+
+  // following instructions mainly focus on NaN, as riscv does not handle
+  // NaN well with vfwcvt_f_f_v, but the code also works for Inf at the same time.
+  //
+  // construct NaN's in 32 bits from the NaN's in 16 bits,
+  // we need the payloads of non-canonical NaNs to be preserved.
+
+  // widen and sign-extend src data.
+  __ vwadd_vx(dst, src, zr, Assembler::v0_t);
+  // adjust vector type to 2 * SEW.
+  __ vsetvli_helper(T_FLOAT, length, Assembler::m1);
+  __ mv(t0, 0x7f800000);
+  // sign-bit was already set via sign-extension if necessary.
+  __ vsll_vi(dst, dst, 13, Assembler::v0_t);
+  __ vor_vx(dst, dst, t0, Assembler::v0_t);
+
+  __ j(stub.continuation());
+#undef __
+}
+
+// j.l.Float.float16ToFloat
+void C2_MacroAssembler::float16_to_float_v(VectorRegister dst, VectorRegister src, uint length) {
+  auto stub = C2CodeStub::make<VectorRegister, VectorRegister, uint>
+              (dst, src, length, 24, float16_to_float_v_slow_path);
+
```

Comment:
I think we should add one assertion here asserting that `dst` and `src` are differenct vector registers.

Reply:
we could, but it's not necessary, as in `riscv_v.ad` it already has:
```
instruct vconvHF2F(vReg dst, vReg src, vRegMask_V0 v0) %{
  ...
  effect(TEMP_DEF dst, TEMP v0);
```

Reply:
We always do that in practice. An explicit assertion will help people better understand the code when they are solely looking at these assember functions. Otherwise they will have to check the caller side about the constraint on register usage. Also it could happen in the future that the same code might be reused or called from other newly-added code. It will help avoid errors on passing registers when people see such an assertion on entry.